fuzz: add ConstructPubKeyBytes util function

josibake opened the pull request:
In https://github.com/bitcoin/bitcoin/pull/28246 and https://github.com/bitcoin/bitcoin/pull/28122 , we add a `PubKeyDestination` and a `V0SilentPaymentsDestination`. Both of these PRs update `fuzz/util.cpp` and need a way to create well-formed pubkeys. Currently in `fuzz/util.cpp`, we have some logic for creating pubkeys in the multisig data provider. This logic is duplicated in #28246 and duplicated again in #28122. Seems much better to have a `ConstructPubKeyBytes` function that both PRs (and any future work) can reuse.

This PR introduces a function to do this and has the existing code use it. While the purpose is to introduce a utility function, the previous multisig code used `ConsumeIntegralInRange(4, 7)` which would have created some uncompressed pubkeys with the prefix 0x05, which is incorrect (see https://bitcoin.stackexchange.com/questions/57855/c-secp256k1-what-do-prefixes-0x06-and-0x07-in-an-uncompressed-public-key-signif)

tldr; using `PickValueFromArray` is more correct as it limits to the set of defined prefixes for compressed and uncompressed pubkeys.

josibake committed:
fuzz: add ConstructPubKeyBytes function

Today, this code only has one spot where it needs well-formed pubkeys,
but future PRs will want to reuse this code.

Add a function which creates a well-formed byte array that can be turned
into a pubkey. It is not required that the pubkey is valid, just that it
can be recognized as a compressed or uncompressed pubkey.

Note: while the main intent of this commit is to wrap the existing
logic into a function, it also switches to `PickValueFromArray` so that
we are only choosing one of 0x04, 0x06, or 0x07. The previous code,
`ConsumeIntegralInRange` would have also picked 0x05, which is not
definied in the context of compressed vs uncompressed keys.

See https://bitcoin.stackexchange.com/questions/57855/c-secp256k1-what-do-prefixes-0x06-and-0x07-in-an-uncompressed-public-key-signif
for more details.

Sjors commented on src/test/fuzz/util.cpp:
```diff
@@ -14,6 +14,20 @@
+    }
+    std::vector<uint8_t> pk_data = ConsumeFixedLengthByteVector(fuzzed_data_provider, pk_type >= 0x04 ? 65 : 33);
+    pk_data[0] = pk_type;
+    CPubKey pk{pk_data.begin(), pk_data.end()};
```
Maybe add a comment to this helper function that the resulting key is not necessarily valid. Or add a `const bool valid` argument which loops until `IsValid()` passes.

josibake replied:
I don't think we need / want strictly valid pubkeys in a fuzzing context. The point of this function is to return "well-formed" but not necessarily valid pubkeys. By well-formed, I mean they have some sort of parity byte, indicating even or oddness, which also indicates the type of pubkey: compressed, uncompressed, hybrid.

Sjors commented:
> Now they will consume bytes for their whole length, which may exhaust the available bytes faster and thus result in all-zeros and also bloat the fuzz inputs git folder

Ah, that's not good.

Sjors commented on src/test/fuzz/util.cpp:
```diff
@@ -103,16 +117,9 @@ CScript ConsumeScript(FuzzedDataProvider& fuzzed_data_provider, const bool maybe
-                            pubkey.back() = num_data; // Make each pubkey different
-                        }
-                        r_script << pubkey;
+                        const auto pubkey = ConsumePubKey(fuzzed_data_provider, fuzzed_data_provider.ConsumeBool());
```
Given @MarcoFalke's point about having to be frugal with consuming input bytes, maybe move this out of the loop and then have the loop tweak the public key. Like before it would start with one compressed and one uncompressed key. We'll still get diversity in `{0x04, 0x06, 0x07}` and `{0x02, 0x03}` combinations between multiple runs.

josibake replied:
I've updated it so that the function now returns a byte vector (which can be mutated if need be) and made it so the function can read from the buffer (as in the multisig example), or be passed some random byte data by calling the `Consume{Random,Fixed}LengthByteVector` functions

josibake commented:
> Not sure why this is called a refactor, when everything is changed.

That was an oversite on my part. I've updated the function to use the buffer as before, so the first commit should be just a refactor now. I also split out adding the missing hybrid pubkey byte into it's own commit.

josibake commented:
> Maybe add 0x07 in a separate commit.
>
> There's another behavior change here, probably a good one: while generating N keys, they're now completely different. > Previously we would only change the last byte to make them unique.

Done! The first commit should now be strictly a refactor.

hebasto commented:
> CI failure seems unrelated to the changes here unless I'm missing something. @hebasto ?

Restarted.

josibake commented:
> > CI failure seems unrelated to the changes here unless I'm missing something. @hebasto ?
> 
> Restarted.

Still failing but with a different functional test failure now, which seems to be related to timeouts.

Sjors commented:
It's green now.

maflcko reviewed (approved):
lgtm, all good.

Though, `refactor` is still wrong, because the fuzz input format is changed.

Generally I don't think the `refactor` label makes any sense for (unit) tests, because end users in production are never affected by test-only changes.

josibake commented on src/test/fuzz/util.cpp:
```diff
@@ -14,6 +14,19 @@
 
 #include <memory>
 
+std::vector<uint8_t> ConstructPubKeyBytes(FuzzedDataProvider& fuzzed_data_provider, const std::vector<uint8_t>& byte_data, const bool compressed) noexcept
```
Updated! Sorry, misread your first comment as having a preference for `const bla&` over `Span<const bla>`. I agree span makes more sense here

josibake commented:
> Though, `refactor` is still wrong, because the fuzz input format is changed.

I removed it from the title, but feel that it's still useful on the commit message. Using the refactor label, to me, is about communicating intent to the reviewer. The first commit in this PR was not meant to change behavior, and calling it a refactor made it more obvious to reviewers that some behavior had in fact changed.

Adding `0x6` is definitely not a refactor (tho still a useful change, imo), which is why I've removed refactor from the title.

josibake commented on src/test/fuzz/util.cpp:
```diff
@@ -14,6 +14,19 @@
+    if (compressed) {
+        pk_type = fuzzed_data_provider.PickValueInArray({0x02, 0x03});
+    } else {
+        pk_type = fuzzed_data_provider.PickValueInArray({0x04, 0x07});
```
yeah, reading again more carefully, the code before was doing `ConsumeIntegralInRange(4, 7)`, which would have included 5 and 6, where 5 is nonsense in this context.

So probably fine to just squash the two commits and mention that it's not strictly a refactor because we are removing a nonsense value from the input.

Sjors replied:
Removing nonsense in the first commit is fine. But I still prefer the keep the second commit seperate.

josibake replied:
Updated the comment!

josibake commented on src/test/fuzz/util.cpp:
```diff
@@ -14,6 +14,19 @@
 
 #include <memory>
 
+std::vector<uint8_t> ConstructPubKeyBytes(FuzzedDataProvider& fuzzed_data_provider, Span<const uint8_t> byte_data, const bool compressed) noexcept
```
I prefer `byte_data` because we are reading in an array of bytes and turning it into `pk_data` by adding a prefix byte to indicate compressed or uncompressed. At the moment we read it in, it is not `pk_data`. We turn it into `pk_data` (i.e. a byte array with the correct prefix), and then that is returned and can be serialized or turned into a `CPubKey` object.

maflcko reviewed (approved):
Seems fine to remove `5`. The reason to have it was to also create "invalid" data and see if it causes a crash. But no strong opinion, as it may be covered by other fuzz tests already.

josibake commented:
> The reason to have it was to also create "invalid" data and see if it causes a crash

That doesn't match the comment, `// Set first byte for GetLen() to pass`, so if that was the intention for this specific test case, we should probably add a step that randomly mutates the bytes by adding an invalid prefix or just reads from the byte buffer without using the function. 

Going forward, I think it's better to have a function just for creating well-formed pubkey data (i.e. correct prefixes) and then use random or invalid data for cases where we actually want to test for that.

Sjors commented:
> Going forward, I think it's better to have a function just for creating well-formed pubkey data (i.e. correct prefixes) and then use random or invalid data for cases where we actually want to test for that.

That makes sense, but since we previously covered `0x05`, the `ConsumeScript` function should occasionally set `0x05` (or some other invalid value) as the prefix.

josibake commented:
> That makes sense, but since we previously covered 0x05 (if only by accident), the ConsumeScript function should occasionally set 0x05 (or some other invalid value) as the prefix.

I'm inclined to leave this pull as is, and if we find there is a real need for mutating with an invalid prefix inside the "push multisig" case, we address it in a follow-up. The way things are written now seems more clear to me: we have a case for sending random bytes to `r_script` and we have a case for creating a well-formed but not necessarily valid multisig script. This seems to better match the intent of the `ConsumeScript` function, based on the comments and surrounding code.

josibake commented:
@MarcoFalke @Sjors can you let me know if this looks good as-is (where some things can be addressed in follow-ups), or if you'd like to see more changes in this PR before merging?

josibake commented:
ping @Sjors @MarcoFalke

Sjors commented:
ACK 1580e3be83bd03985b2f288ed70de510903068d9
